src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs

```diff
@@ -282,27 +283,110 @@ public bool IsPlatformNeutral
+            TypeDefinitionHandle[] bucketHeads = new TypeDefinitionHandle[(buckets.Length / 8) + 1];
+
+            MetadataReader reader = _metadataReader;
+            foreach (TypeDefinitionHandle typeHandle in reader.TypeDefinitions)
+            {
+                TypeDefinition typeDef = reader.GetTypeDefinition(typeHandle);
+                if (typeDef.Attributes.IsNested())
+                    continue;
+
+                var hashCode = default(HashCode);
+                AppendHashCode(ref hashCode, reader.GetBlobReader(typeDef.Namespace));
+                AppendHashCode(ref hashCode, reader.GetBlobReader(typeDef.Name));
+
+                ref TypeDefinitionHandle head = ref bucketHeads[(uint)hashCode.ToHashCode() % bucketHeads.Length];
+                ref TypeDefinitionHandle entry = ref buckets[MetadataTokens.GetRowNumber(typeHandle)];
+
+                entry = head;
+                head = typeHandle;
+            }
+
+            _typeDefinitionBuckets = buckets;
+            _typeDefinitionBucketHeads = bucketHeads;
+
+            return bucketHeads;
+
+            static unsafe void AppendHashCode(ref HashCode hashCode, BlobReader reader)
+            {
+                while (reader.RemainingBytes > 0)
+                    hashCode.Add(reader.ReadByte());
+            }
+        }
+
+        private TypeDefinitionHandle FindType(string nameSpace, string name)
+        {
+            var hashCodeBuilder = default(HashCode);
+            AppendHashCode(ref hashCodeBuilder, nameSpace);
+            AppendHashCode(ref hashCodeBuilder, name);
+
+            MetadataReader reader = _metadataReader;
+            MetadataStringComparer stringComparer = reader.StringComparer;
+
+            TypeDefinitionHandle[] bucketHeads = _typeDefinitionBucketHeads ?? InitializeFindNamedType();
+            TypeDefinitionHandle entry = bucketHeads[(uint)hashCodeBuilder.ToHashCode() % bucketHeads.Length];
+            while (!entry.IsNil)
+            {
+                var typeDefinition = reader.GetTypeDefinition(entry);
+                if (stringComparer.Equals(typeDefinition.Name, name) &&
+                    stringComparer.Equals(typeDefinition.Namespace, nameSpace))
+                {
+                    return entry;
+                }
+
+                entry = _typeDefinitionBuckets[MetadataTokens.GetRowNumber(entry)];
+            }
+
+            return default;
+
+            static void AppendHashCodeNonAscii(ref HashCode hashCode, ReadOnlySpan<char> s)
+            {
+                const int MaxStackAlloc = 1024;
+                int maxByteCount = Encoding.UTF8.GetMaxByteCount(s.Length);
+                Span<byte> utf8 = maxByteCount <= MaxStackAlloc ? stackalloc byte[MaxStackAlloc] : new byte[maxByteCount];
+                int actualBytes = Encoding.UTF8.GetBytes(s, utf8);
+
+                foreach (byte b in utf8.Slice(0, actualBytes))
```

Comment:
AddBytes(..) ?

Reply:
It comes up with different numbers than calling Add on individual bytes. Already got burned by it.

Reply:
Why does that matter in this case? As long as the type consistently does the same thing on each overall hash code computation, what does this negatively impact?

Reply:
> This code wanted to efficiently compute the same hash code from both UTF8-based strings and Unicode-based strings.

Ah, ok.